enterprise/tpa_pipeline.py

```diff
@@ -0,0 +1,153 @@
+"""
+Module provides elements to be used in third-party auth pipeline
+"""
+from django.core.urlresolvers import reverse
+from django.shortcuts import redirect
+from django.utils.translation import ugettext as _
+
+try:
+    from third_party_auth.provider import Registry
+except ImportError:
+    Registry = None
+
+from enterprise.models import EnterpriseCustomer, UserDataSharingConsentAudit
+from enterprise.utils import NotConnectedToEdX
+
+from social.pipeline import partial
+
+class MultipleCustomersFound(Exception):
+    """
+    Raise this exception if there are multiple customers tied to an IdP.
+    """
+
+
+def get_enterprise_customer_for_request(request):
+    """
+    Get the EnterpriseCustomer associated with a particular request.
+    """
+    if Registry is None:
+        raise NotConnectedToEdX(
+            _("This package must be installed in an EdX environment to look up third-party auth providers.")
+        )
+    pipeline = request.session.get('partial_pipeline')
+    provider = Registry.get_from_pipeline(pipeline)
+    return get_enterprise_customer_for_sso(provider)
+
+
+def get_enterprise_customer_for_running_pipeline(pipeline):
+    """
+    Get the EnterpriseCustomer associated with a running pipeline
+    """
+    if Registry is None:
+        raise NotConnectedToEdX(
+            _("This package must be installed in an EdX environment to look up third-party auth providers.")
+        )
+
+    provider = Registry.get_from_pipeline(pipeline)
+    return get_enterprise_customer_for_sso(provider)
+
+
+def get_enterprise_customer_for_sso(provider):
+    """
+    Get the EnterpriseCustomer object tied to an identity provider.
+    """
+    if provider is None:
+        return None
+    try:
+        return EnterpriseCustomer.objects.get(identity_provider=provider.idp_slug)
+    except EnterpriseCustomer.MultipleObjectsReturned:
+        raise MultipleCustomersFound(
+            _("Multiple EnterpriseCustomer objects are tied to the same identity provider; expected one.")
+        )
+    except EnterpriseCustomer.DoesNotExist:
+        return None
+
+
+def active_provider_enforces_data_sharing(request, enforcement_location):
+    """
+    Determine two things - first, whether there's an active third-party
+    identity provider currently running, and second, if that active provider
+    enforces data sharing consent at the given point in order to proceed.
+
+    Args:
+        request: HttpRequest object containing request data.
+        enforcement_location (str): the point where to see data sharing consent state.
+        argument can either be "optional", 'at_login' or 'at_enrollment'
+    """
+    running_pipeline = request.session.get('partial_pipeline')
+    if running_pipeline:
+        customer = get_enterprise_customer_for_request(request)
+        return customer and customer.enforces_data_sharing_consent(enforcement_location)
+    return False
+
+
+def active_provider_requests_data_sharing(request):
```

Comment:
@mattdrayer In that case, perhaps the SSO code in the platform would use an API to ask the enterprise service if the provider requests data sharing, and then the enterprise service could requests that specific users  be redirected to the enterprise service (at a separate URL?) to consent during the SSO flow.

I'm still a bit fuzzy on how this thing will work when it becomes a service or even just gets used as a library by multiple IDAs. I think we should put more thought into that now.